pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py

```diff
@@ -244,7 +252,19 @@ def _blocking_consume(self):
                 break
             except Exception as exc:
                 recover = self._policy.on_exception(exc)
-                if not recover:
+                if recover:
+                    with threading.Lock():
+                        # Must lock so that ``self.send_request()`` cannot add
+                        # more items to the queue while it is being replaced.
+                        previous_queue = self._request_queue
```

Comment:
@lukesneeringer After chatting with @jonparrott it seems there is another option here.

If we could guarantee that `self._request_queue.empty()` is `True`, then we could just send `STOP` into the queue and wait until the `STOP` gets consumed, then we don't have to swap the queue.

(As you can see from the coverage failure, the queue will always be empty.)

Reply:
That makes sense to me.

Reply:
#4504 is an implementation of that other option